source/MRViewer/MRSurfaceContoursWidget.cpp

```diff
@@ -0,0 +1,421 @@
+            assert( pickedIndex != pickedPoints_[pickedObj].size() - 1 ); // unable to pick point which is close countour
+
+            SCOPED_HISTORY( "Remove Point" );
+
+            // 4 points - minimal non-trivial closed path
+            // last on is a "pseudo" point to close contour
+            // so remove last point which is close contour in case of: 
+            // 1) Contour can no longer been closed (only 2 valid point left ) 
+            // 2) First point need to be remove. Last point will be restored later.  
+            if ( pickedPoints_[pickedObj].size() == 4 || pickedIndex == 0 )
+                removePoint( pickedObj, ( int )pickedPoints_[pickedObj].size() - 1 );
+
+            // Remove point marked to be removed. 
+            removePoint( pickedObj, pickedIndex );
+
+            // Restore close countour. 
+            if ( pickedPoints_[pickedObj].size() > 2 && pickedIndex == 0 )
+                addPoint( pickedObj, pickedPoints_[pickedObj][0]->getCurrentPosition(), true );
+        }
+        else
+            removePoint( pickedObj, pickedIndex );
+    }
+    return false;
+}
+
+bool SurfaceContoursWidget::onMouseMove_( int, int )
+{
+    if ( !isPickerActive_ )
+        return false;
+
+    if ( pickedPoints_.empty() || activeChange_ )
+        return false;
+
+    auto [obj, pick] = getViewerInstance().viewport().pick_render_object();
+    if ( !obj )
+        return false;
+
+    for ( auto contour : pickedPoints_ )
+        for ( int i = 0; i < contour.second.size(); ++i )
+        {
+            const auto& point = contour.second[i];
+            bool hovered = obj == point->getPickSphere();
+            point->setHovered( hovered );
+            if ( hovered )
+            {
+                activeIndex_ = i;
+                activeObject_ = contour.first;
+            }
+        }
+    return false;
+}
+
+void SurfaceContoursWidget::create( PickerPointCallBack onPointAdd, PickerPointCallBack onPointMove, PickerPointCallBack onPointMoveFinish, PickerPointCallBack onPointRemove )
+{
+    onPointAdd_ = std::move( onPointAdd );
+    onPointMove_ = std::move( onPointMove );
+    onPointMoveFinish_ = std::move( onPointMoveFinish );
+    onPointRemove_ = std::move( onPointRemove );
+
+    clear();
+
+    // 10 group to imitate plugins behavior
+    connect( &getViewerInstance(), 10, boost::signals2::at_front );
+}
+
+void SurfaceContoursWidget::clear()
+{
+    pickedPoints_.clear();
+    activeIndex_ = 0;
+    activeObject_ = nullptr;
+}
+
+void SurfaceContoursWidget::reset()
+{
+    clear();
+    enable( false );
+
+    FilterHistoryByCondition( [&] ( const std::shared_ptr<HistoryAction>& action )
+    {
+        bool res = bool( std::dynamic_pointer_cast< AddPointActionPickerPoint >( action ) ) ||
+            bool( std::dynamic_pointer_cast< RemovePointActionPickerPoint >( action ) ) ||
+            bool( std::dynamic_pointer_cast< ChangePointActionPickerPoint >( action ) );
+        return res;
+    } );
```

Comment:
widget should not manipulate with history by it-self, history should be written and filtered only by user in plugin